domain/tutoringModel/Tyche/tyche_utils.py

```diff
@@ -0,0 +1,330 @@
+
+        # Insert missing probabilities for manuscript:
+        if counter_ms > 1:
+            duplicate_column = final_prob[:, counter_ms]
+            for _i in range(counter_ms - 1):
+                final_prob = np.insert(
+                    final_prob, counter_ms + 1, duplicate_column, axis=1
+                )
+    return final_prob
+
+
+def add_qu_probs(flag_quiz, final_prob, les):
+    index_qu1 = False
+    index_qu2 = False
+    first_qu_prob = -1
+    if flag_quiz:
+        counter_qu = 0
+        if "quiz_rq" in les:
+            counter_qu = counter_qu + 1
+            index_qu1 = les.index("quiz_rq")
+            if first_qu_prob > index_qu1:
+                first_qu_prob = index_qu1
+        if "quiz_se" in les:
+            counter_qu = counter_qu + 1
+            index_qu2 = les.index("quiz_se")
+            if first_qu_prob > index_qu2:
+                first_qu_prob = index_qu2
+
+        # Insert missing probabilities for quiz:
+        if counter_qu == 2:
+            duplicate_column = final_prob[:, counter_qu]
+            final_prob = np.insert(final_prob, counter_qu + 1, duplicate_column, axis=1)
+    return final_prob
+
+
+def add_animation_probs(flag_animation, final_prob, les, le_weight):
+    if flag_animation:  # weighted average of AAM and VAM
+        # Get probs of AAM and VAM
+        colum_sum = []
+        if "zusatzmaterial_textuell" in les:
+            w = 3
+        else:
+            w = 2
+        for i in range(len(final_prob)):
+            le_t = [final_prob[i][-1], final_prob[i][-w]]
+            colum_sum.append([np.average(le_t, weights=le_weight)])
+
+        # Delete probs of AAM and VAM in final_prob
+        final_prob = np.delete(final_prob, -1, axis=1)
+        final_prob = np.delete(final_prob, ((w - 1) * (-1)), axis=1)
+
+        # Add column of animation at the end of final_prob:
+        final_prob = np.append(final_prob, colum_sum, axis=1)
+        pass
+    return final_prob
+
+
+def get_probability_columns(les, le_weight, final_prob):
+    final_prob = np.asarray(final_prob)
+
+    flag_manuskript, flag_quiz, flag_animation, final_prob = set_le_flags(
+        les, final_prob
+    )
+
+    # Adapt probabilities for animation:
+    final_prob = add_animation_probs(flag_animation, final_prob, les, le_weight)
+
+    # Adapt probabilities for manuskript:
+    final_prob = add_ms_probs(flag_manuskript, final_prob, les)
+
+    # Adapt probabilities for quiz:
+    final_prob = add_qu_probs(flag_quiz, final_prob, les)
+    return final_prob
+
+
+def get_probability_matrix(probability, les):
+    le_weight = [0.75, 0.25]
+    les_len = len(les)
+
+    # Transform probability matrix:
+    final_prob = get_probability_rows(probability, les, le_weight)
+
+    # Get columns:
+    final_prob = get_probability_columns(les, le_weight, final_prob)
+
+    # Check probability matrix:
+    if not check_probability_matrix(les_len, final_prob):
+        print("Error in matrix generation")

```

Comment:
If actual error use err.ErrorException. Depends if the execution flow should be interrupted or not.

Reply:
is changed to err.ErrorException(...)